src/cr/cube/slices.py

```diff
@@ -0,0 +1,1140 @@
+        )
+
+    @lazyproperty
+    def intersections(self):
+        intersections = []
+        for row in self._rows:
+            row_col_intersections = []
+            row_idx = (
+                0
+                if row.anchor == "top"
+                else -1
+                if row.anchor == "bottom"
+                else row.anchor
+            )
+            for col in self._columns:
+                col_idx = (
+                    0
+                    if col.anchor == "top"
+                    else -1
+                    if col.anchor == "bottom"
+                    else col.anchor + 1
+                )
+                intersection_value = row.values[col_idx] + col.values[row_idx]
+                row_col_intersections.append(intersection_value)
+            intersections.append(row_col_intersections)
+
+        return np.array(intersections)
+
+
+class _InsertionVector(object):
+    def __init__(self, slice_, subtotal):
+        self._slice = slice_
+        self._subtotal = subtotal
+
+    @lazyproperty
+    def table_margin(self):
+        return self._slice.table_margin
+
+    @lazyproperty
+    def anchor(self):
+        return self._subtotal.anchor_idx
+
+    @lazyproperty
+    def addend_idxs(self):
+        return np.array(self._subtotal.addend_idxs)
+
+    @lazyproperty
+    def values(self):
+        return np.sum(np.array([row.values for row in self._addend_vectors]), axis=0)
+
+    @lazyproperty
+    def base_values(self):
+        return np.sum(
+            np.array([row.base_values for row in self._addend_vectors]), axis=0
+        )
+
+    @lazyproperty
+    def margin(self):
+        return np.sum(np.array([vec.margin for vec in self._addend_vectors]), axis=0)
+
+    @lazyproperty
+    def base(self):
+        return np.sum(np.array([vec.base for vec in self._addend_vectors]), axis=0)
+
+    @lazyproperty
+    def proportions(self):
+        return self.values / self.margin
+        # return self.values / self.base
+
+    @lazyproperty
+    def table_proportions(self):
+        return self.values / self.table_proportions
+
+    @lazyproperty
+    def pruned(self):
+        """Insertions are never pruned."""
+        return False
+
+
+class InsertionRow(_InsertionVector):
+    @lazyproperty
+    def _addend_vectors(self):
+        return tuple(
+            row
+            for i, row in enumerate(self._slice.rows)
+            if i in self._subtotal.addend_idxs
+        )
+
+
+class InsertionColumn(_InsertionVector):
+    @lazyproperty
+    def _addend_vectors(self):
+        return tuple(
+            row
+            for i, row in enumerate(self._slice.columns)
+            if i in self._subtotal.addend_idxs
+        )
+
+
+class Assembler(object):
+    def __init__(self, slice_, transforms):
+        self._slice = slice_
+        self._transforms = transforms
+
+    @lazyproperty
+    def slice(self):
+        """Apply all transforms sequentially."""
```

Comment:
This will need to apply the transforms sequentially. It might even turn out to be agnostic of the application order, but for now I'd keep it rigid.